README.md
# Gengine-CSC481_591
Game engine developement through game creation as part of CSC481_591 project.

Needs to be run in Google Chrome




GameEngine/ChangeLog.txt
Snake Refactoring - 9/22/2017
Commit Messages on github show the specific changes in our code.
 https://github.com/smscoggi/Gengine-CSC481_591/commits/master/GameEngine

- Moved gameloop() and setInterval() to the engine
:github commit = gameloop, makefoodcollision

- Moved var localHighscore to the engine. Created a function, checkHighscore(score).
Takes score and checks if it is greater than localHighscore, if yes then store it.
:github commit = A couple of changes

- Fixed food reset on snake's death bug. Now food properly resets when snake dies.
Orginally food did not reset on snake's death.
:github commit = A couple of changes
 
- Moved makefood() to engine. Made the function more general to apply to multiple games.
All related varibles to makefood() moved to array.
:github commit = moved makefood to engine

- Moved grid creation and highscore variables to the engine.
:github commit = gridfunction,highscore variables, random

- Put direction of snake movement in engine.
:github commit = took out snake movement

- Added changing difficulty for snake game.
:github commit = diffuculty implementation
:github commit = Fixed Difficulty Dropdown

Asteroids Refactoring - 11/1/2017
-Added a menu. Press Esc to start/pause.

-Changed Makebullet() to reflect pause menu.

-Moved various functions from Asteroids.js to Engine.js
	-Functions
	:adjustPosXYByCenterPoint
	:adjustXYByCenterPoint
	:rotatingDirection
	:rightRotation
	:leftRotation
	:angledFowardMotion
	:jumpToOtherSideOfScreen
	:testSideCrossing
	
-Comment cleanup and blank space cleanup

-BUG fix - Added Asteroids to gallery in GamesIndex.html

-Took out update velocity from drawSprites and made it to its own function.

-BUg fix - Flying bullets are now destroyed when game resets.

-Added border to score box

Snake Multiplayer
-Added Snake Local multiplayer
-BUG Fix - apples no longer spawn inside walls.

-BUG Fix - switching from single to multiplayer mode bug fixed. Functions properly
	pause the game with 'Esc' and switch game mode if wanted.

-Cops and robbers Added

Astar search algorithm added to Engine with heapify,road, tile, path functions

-Converted MakeAsteroid function into a MakeParticles function for generalized particle use. Asteroid specific functionality taken out 
	to another function. MakeParticles allows for randomly moving sprites, of varying size, speed, and rotation, allows for spawn distance
	from a specified point. All particle attributes that have asteroid presets have flexibility to manipulate them for game specific content.

-refactored menu function and draw menu moved to engine to hold menu details specified by game.

-refactored peer to peer code in snake.js for more generalization and moved relevant functions to engine. Developer can manipulate functions
	called on connections to fit game specific needs


GameEngine/Snake.js
//peer to peer

peer1OnStartConnection= function(data){
	drawnSpritesConnectUpdate(data);
	snakeConnectUpdate(data);
	directionConnectUpdate(data);
	scoringConnectUpdate(data);
	menuConnectUpdate(data);

}
peer1OnGetConnection=function(data){
	directionConnectUpdate(data);
	menuConnectUpdate(data);
}
function snakeConnectUpdate(data){
	if(data.constructor === Array){
		if(data[0]=="sn"){
			//console.log("snakkke"+ data[6]);
			if(data[4]==1){
				//snake1.numLinks=data[1];
				//snake1.posX= data[2];
				//snake1.posY= data[3];
				//snake1.deadState=data[5];
				
				snake1.snakeArray=data[6];
				
				//console.log(data[6].length);
			}
			else{
				snake2.snakeArray=data[6];
			}


	}}
}


/////Snake
var difficulty = 2;

var highscore2=0;
var score2=0;

var player1XPos = Math.floor(xcellCount/4)*1-1;
var player1YPos = Math.floor(ycellCount/2)-1;
var player2XPos = Math.floor(xcellCount/4)*3-1;
var player2YPos = Math.floor(ycellCount/2)-1;

var snake1 = new Snake(5, player1XPos, player1YPos, 1);
var snake2 = new Snake(5, player2XPos, player2YPos, 2);

function newSnake(snakeID){
	if (snakeID == 1) {
		snake1 = new Snake(5, player1XPos, player1YPos, 1);
	}
	if (snakeID == 2) {
		snake2 = new Snake(5, player2XPos, player2YPos, 2);
	}
}

function Snake(numLinks, posX, posY, snakeID){
	var segPos = 0;
	var backwardsDirection = 'none';
	this.posX = posX;
	this.posY = posY;
	this.snakeLinks = numLinks;
	this.snakeArray = new Array();
	this.deadState = 'notDead';
	this.snakeID = snakeID;

	this.update = function(){
		if(!connected2){
		if(snakeID == 1){
			backwardsDirection=basicDirection(this, direction,backwardsDirection);
		}
		if(snakeID == 2){
			backwardsDirection=basicDirection(this, direction2,backwardsDirection);
		}
		var tempObject =jumpToOtherSideOfScreen(this);
		this.posX= tempObject.posX;
		this.posY = tempObject.posY;
		

		this.snakeArray.push({x:this.posX, y:this.posY, sP:segPos});
		segPos = 0;
		for(var iter = 0; iter < this.snakeArray.length; iter++) {
			this.snakeArray[iter].sP = segPos++;
		}
		while(this.snakeArray.length > this.snakeLinks) {
			this.snakeArray.shift();
		}

		for(var iter = 0; iter < this.snakeArray.length-1; iter++) {
			if(snakeID == 1){
				if (checkdistance(this.snakeArray[iter].x, this.snakeArray[iter].y, this.posX, this.posY, 0) && direction != 'none') {
					this.deadState = 'dead';
				}
				for(var jter = 0; jter < snake2.snakeArray.length-1; jter++) {
					if (checkdistance(this.snakeArray[iter].x, this.snakeArray[iter].y, snake2.snakeArray[jter].x, snake2.snakeArray[jter].y, 0) && direction != 'none') {
						this.deadState = 'dead';
					}
				}
				
			}
			if(snakeID == 2){
				if (checkdistance(this.snakeArray[iter].x, this.snakeArray[iter].y, this.posX, this.posY, 0) && direction2 != 'none') {
					this.deadState = 'dead';
				}
				for(var jter = 0; jter < snake1.snakeArray.length-1; jter++) {
					if (checkdistance(this.snakeArray[iter].x, this.snakeArray[iter].y, snake1.snakeArray[jter].x, snake1.snakeArray[jter].y, 0) && direction2 != 'none') {
						this.deadState = 'dead';
					}
				}
			}
		}

		if (this.deadState == 'dead') {
			newSnake(this.snakeID);
		
			if(this.snakeID == 1){
				direction = 'none';
				score = 0;
			}
			if(this.snakeID == 2){
				direction2 = 'none';
				score2 = 0;

				if(connected1){
					conn.send(["dir",'none']);
				}
			}
			
			

			//resetFood();

			this.deadState = 'notDead';
		}
		//console.log(this.snakeArray[0].x, posX);
/////test to see if snake ran into food/////

		for(var iter=0; iter<foodArray.length; iter++){
			foodCollide=foodArray[iter].collision(this.posX,this.posY,0);

			if (foodCollide == 1) { 
				this.snakeLinks = this.snakeLinks - 1;
				if (this.snakeLinks < 2) {
					this.deadState = 'dead';
				}
				if(this.snakeID==2){
					score2--;
				}
				else {score--; }
				foodArray[iter].reset();
			}
			else if (foodCollide==2) { 
				this.snakeLinks++;
				if(this.snakeID==2){
					score2++;
				}
				else{score++;}
				foodArray[iter].reset();
			}

		}

	}
	};
	this.draw = function(){
		
		if(this.snakeID == 1) {
			for(var iter=0; iter<this.snakeArray.length; iter++){
				context.fillStyle='green';
				context.fillRect(this.snakeArray[iter].x*cellWidth, this.snakeArray[iter].y*cellHeight, cellWidth, cellHeight);
				context.strokeStyle = 'yellow';
				context.strokeRect(cellWidth*this.snakeArray[iter].x, this.snakeArray[iter].y*cellHeight, cellWidth, cellHeight)
			}
		}
		
		if(this.snakeID == 2) {
			for(var iter=0; iter<this.snakeArray.length; iter++){
				context.fillStyle='blue';
				context.fillRect(this.snakeArray[iter].x*cellWidth, this.snakeArray[iter].y*cellHeight, cellWidth, cellHeight);
				context.strokeStyle = 'yellow';
				context.strokeRect(cellWidth*this.snakeArray[iter].x, this.snakeArray[iter].y*cellHeight, cellWidth, cellHeight)
			}
		}

	};



} //Snake


//Food Objects
var amountFood = 3;
var foodMaxLifeTime = 100;
var foodSpoilTime = 65;

addSprite("https://i.imgur.com/2IhGuoJ.png","redapple");
addSprite("https://i.imgur.com/dIPgiDN.png","greenapple");
var redApple = findSprite(sprites,"redapple");
var greenApple = findSprite(sprites,"greenapple");

makeFood(amountFood,foodSpoilTime,foodMaxLifeTime,redApple,greenApple);
//Food Objects


//Wall Objects
addSprite("https://i.imgur.com/3vugqUb.jpg","walltexture");
var wallSprite = findSprite(sprites,"walltexture");
var wallArray = new Array();
wallArray.push(new Wall(0,0,3,1));
wallArray.push(new Wall(6,4,4,1));
wallArray.push(new Wall(20,25,1,3));
wallArray.push(new Wall(2,24,5,1));
wallArray.push(new Wall(24,2,1,4));

for(var iter = 0; iter < wallArray.length; iter++){

	addDrawnSprites(wallSprite,wallArray[iter].xWall*cellWidth, wallArray[iter].yWall*cellHeight, cellWidth*wallArray[iter].xcellWidth, cellHeight*wallArray[iter].ycellLength,wallSprite.ID);

	/*wallArray[iter].draw = function wallDraw(xWall, yWall, xcellWidth, ycellLength) {
			context.fillStyle='blue';
			context.fillRect(xWall*cellWidth, yWall*cellHeight, cellWidth*xcellWidth, cellHeight*ycellLength);
		
	}*//////////// for block walls
} 


function makeWalls(){
	for(var iter = 0; iter < wallArray.length; iter++) {
		var wallid = "w"+iter;
		var wall = new Wall()
		var ThisWallSprite;

		for (var jter = 0; jter < wallArray[iter].xcellWidth; jter++) {
			for (var kter = 0; kter < wallArray[iter].ycellLength; kter++) {
				if(singlePlayerMode){
					if (checkdistance(wallArray[iter].xWall+jter, wallArray[iter].yWall+kter, snake1.posX, snake1.posY, 0)) { 	
						snake1.deadState = 'dead';
					}
				}
				
				if(multiPlayerMode){
					if (checkdistance(wallArray[iter].xWall+jter, wallArray[iter].yWall+kter, snake1.posX, snake1.posY, 0)) { 	
						snake1.deadState = 'dead';
					}
					if (checkdistance(wallArray[iter].xWall+jter, wallArray[iter].yWall+kter, snake2.posX, snake2.posY, 0)) { 	
						snake2.deadState = 'dead';
					}
				}
				
				for(var lter = 0; lter < foodArray.length; lter++) {
					if (checkdistance(wallArray[iter].xWall+jter, wallArray[iter].yWall+kter, foodArray[lter].xfood, foodArray[lter].yfood, 0)) { 	
						foodArray[lter].reset();
					}
				}
				
			}
		}

	}
}
//Wall Objects






function drawStats(){
	
		var textfont ="verdana";
		var textfillstyle = "#FFFFFF";
		var textsize = 10;
	
		var score_text = "Player1 Score: " + score ;
			//context.fillStyle="#FFFFFF";
			//context.fillText(score_text, 5, canvas.height-5);
		addText(textsize,textfont,score_text,5, canvas.height-5,textfillstyle);
	
		var highscore_text = "Player1 Highscore: " + highscore ;
			//context.fillStyle="#FFFFFF";
			//context.fillText(highscore_text, 5, canvas.height-20);
		addText(textsize,textfont,highscore_text,5, canvas.height-20,textfillstyle);

		if(multiPlayerMode){
			var score_text = "Player2 Score: " + score2 ;
			//context.fillStyle="#FFFFFF";
			//context.fillText(score_text, 5, canvas.height-5);
		addText(textsize,textfont,score_text,150, canvas.height-5,textfillstyle);
	
		var highscore_text = "Player2 Highscore: " + highscore2 ;
			//context.fillStyle="#FFFFFF";
			//context.fillText(highscore_text, 5, canvas.height-20);
		addText(textsize,textfont,highscore_text,150, canvas.height-20,textfillstyle);
		}
	
	
		var highscore_text2 = "Alltime Highscore: " + localhighscore;
			//context.fillStyle="#FFFFFF";
			//context.fillText(highscore_text2, 5, canvas.height-45);
		addText(textsize,textfont,highscore_text2,5, canvas.height-45,textfillstyle);
		
	
	}

function setDifficulty(difficulty){
	if(connected2){
		console.log("Can't change difficulty as client");
		return;
	}
	//difficulty = document.getElementById('difficulty').innerHTML;
	//console.log(difficulty);
	switch(difficulty){
		case 1:
			setGameLoopInterval(110);
			resetFood();
			snake1.deadState = "dead";
			snake2.deadState = "dead";
			break;
		case 2:
			setGameLoopInterval(80);
			resetFood();
			snake1.deadState = "dead";
			snake2.deadState = "dead";
			break;
		case 3:
			setGameLoopInterval(50);
			resetFood();
			snake1.deadState = "dead";
			snake2.deadState = "dead";
			break;
	}
}

function changeDifficulty() {
	var x = document.getElementById("difficulty").options.selectedIndex;
	setDifficulty(x);
}



/////////////////////////////////////////////////////////UPDATE + Draw

///menu stuff
var onMenu = true;
var titleScreen = new Menu('Snake', '48px', 'monospace', 'white', context.canvas.width, context.canvas.height / 2 - 30, 'centered');

var text1 = 'Single Player';
context.font = '24px monospace';
var text1Measurement = context.measureText(text1);
var text1XPos = (context.canvas.width - text1Measurement.width) / 2;
var text1YPos = context.canvas.height / 2 + 30;
var highlightText1 = false;

var text2 = 'Multiplayer';
context.font = '24px monospace';
var text2Measurement = context.measureText(text2);
var text2XPos = (context.canvas.width - text2Measurement.width) / 2;
var text2YPos = context.canvas.height / 2 + 60;
var highlightText2 = false;


var singlePlayerMode = false;
var multiPlayerMode = false;


function leftmouseupSinglePlayer(e) {
	singlePlayerMode = true;
	player1XPos = Math.floor(xcellCount/2)*1-1;
	player1YPos = Math.floor(ycellCount/2)-1;

	//resets game;
	snake1.deadState = "dead";
	multiPlayerMode = false;
	menu++;
}

function leftmouseupMultiplayer(e) {
	peerID2=getOtherPlayer()
	connectToPeer2(peerID2);
	multiplayerSnakeSetup();
}

function multiplayerSnakeSetup(){
	

	multiPlayerMode = true;
	player1XPos = Math.floor(xcellCount/4)*1-1;
	player1YPos = Math.floor(ycellCount/2)-1;
	player2XPos = Math.floor(xcellCount/4)*3-1;
	player2YPos = Math.floor(ycellCount/2)-1;

	//resets game;
	snake1.deadState = "dead";
	snake2.deadState = "dead";
	singlePlayerMode = false;
	menu++;
}


function update() {
	

	if(menu == 0) {
		onMenu = true;
	} else {
		onMenu = false;
	}
	if(connected1){
		conn.send(["menu",menu]);
	}
	
	if(onMenu) {
		if(mouseisdown == 'yes'){
			if(checkdistance(text1XPos + text1Measurement.width / 2, 0, xcoord, 0, text1Measurement.width / 2) &&
			checkdistance(0, text1YPos - 12, 0, ycoord, 12)) {
				highlightText1 = true;
				canvas.addEventListener("mouseup",leftmouseupSinglePlayer);
			}

		}
		
		if(mouseisdown == 'yes'){
			if(checkdistance(text2XPos + text2Measurement.width / 2, 0, xcoord, 0, text2Measurement.width / 2) &&
			checkdistance(0, text2YPos - 12, 0, ycoord, 12)) {
				highlightText2 = true;
				canvas.addEventListener("mouseup",leftmouseupMultiplayer);
			}

		}
		
	}
	
	if(!onMenu){
		if(singlePlayerMode){
			canvas.removeEventListener("mouseup",leftmouseupSinglePlayer);
			snake1.update();
		}
		if(multiPlayerMode){
			
			canvas.removeEventListener("mouseup",leftmouseupMultiplayer);
			
			snake1.update();
			snake2.update();
			//snake3.update();

			

		}
		makeWalls();
		for(var iter = 0; iter < foodArray.length; iter++){
			foodArray[iter].update();
			//foodArray[iter].draw();
			
		}

		if(!connected2){
			
			if(connected1){

				aswd=false;
				//for(var i=0; i<drawnSprites.length; i++){
				//	var dsArray= ["ds",drawnSprites[i].X, drawnSprites[i].Y,drawnSprites[i].image.width, drawnSprites[i].image.height, drawnSprites[i].image.src, drawnSprites[i].ID]
				//	conn.send(dsArray);
				//}

				for(var i=0; i<foodArray.length; i++){
					fsprite=findSprite(drawnSprites,foodArray[i].id);
					//console.log(fsprite);
					var dsArray= ["ds",fsprite.X, fsprite.Y,fsprite.image.width, fsprite.image.height, fsprite.image.src, fsprite.ID]
					conn.send(dsArray);
				}


				var sn1Array=["sn",snake1.numLinks,snake1.posX,snake1.posY,snake1.snakeID,snake1.deadState,snake1.snakeArray];
				var sn2Array=["sn",snake2.numLinks,snake2.posX,snake2.posY,snake2.snakeID,snake2.deadState,snake2.snakeArray];
					conn.send(sn1Array);
					conn.send(sn2Array);	
				
				var statsArray= ["statsA",score,score2,highscore,highscore2,localhighscore];
				conn.send(statsArray);
			}
		}
		if(gotAConnection){
			multiplayerSnakeSetup();
			gotAConnection=false;
		}

		if(connected2){

			arrows=false;
			foodArray.length=0;

			var DirArray=["dir",direction2];
			conn.send(DirArray);

			


		}

		//////////////////////////

		if (highscore < score) {
			highscore = score;
		}
		if(highscore2 < score2){
			highscore2= score2;
		}

		
	}
	
	checkHighscore(score);
	checkHighscore(score2);


}


function draw() {
	canvas.width = canvas.width;
	context.fillStyle="black";
	context.fillRect(0,0,canvas.width,canvas.height);

	//testDrawnSprites();
	//testSprites();

	if(singlePlayerMode){
		snake1.draw();
	}
	if(multiPlayerMode){
		snake1.draw();
		snake2.draw();
		//snake3.draw();
	}
	
	drawSprites();

	drawStats();
	
	if(onMenu){
		drawMenu(titleScreen);
		
		context.fillStyle = 'red';
		context.strokeStyle = 'red';
		if(highlightText1 == true) {
			context.fillStyle = '#ff9999';
			context.strokeStyle = '#ff9999';
			highlightText1 = false;
		}
		addText('24px', 'monospace', text1, text1XPos, text1YPos);
		context.strokeRect(text1XPos, text1YPos, text1Measurement.width, -24);
		
		context.fillStyle = 'green';
		context.strokeStyle = 'green';
		if(highlightText2 == true) {
			context.fillStyle = '#00ff00';
			context.strokeStyle = '#00ff00';
			highlightText2 = false;
		}
		addText('24px', 'monospace', text2, text2XPos, text2YPos);
		context.strokeRect(text2XPos, text2YPos, text2Measurement.width, -24);

		
		var text3 = "To host multiplayer game, "
		if(peer1ID){
			text3 += "have friend join " + peer1ID;
		}
		else{
			text3 += "establish a connection"
		}
		context.font = '12px monospace'
		var text3Measurement = context.measureText(text3);
		var text3XPos = (context.canvas.width - text3Measurement.width) / 2;
		var text3YPos = text2YPos +20;
		context.fillStyle = 'white';
		context.strokestyle = 'white';
		context.font = '12px monospace';
		context.fillText(text3, text3XPos, text3YPos);
		//context.strokeRect(text3XPos, text3YPos, text3Measurement.width, -24);
	} 
}


GameEngine/Cops_Robbers.js

////set up grid 
var CRcellWidth = 35;
var CRcellHeight = 35;
var CRxcellCount = Math.floor(canvas.width/CRcellWidth);
var CRycellCount = Math.floor(canvas.height/CRcellHeight-1);
var gameFinished = 0;
var maxTurns = 30;

setCanvasGrid(CRcellWidth,CRcellHeight, CRxcellCount, CRycellCount);

LevelGridArray=[1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0,
                1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,
                1,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,
                1,0,1,1,1,1,1,1,1,1,0,0,0,1,1,1,1,1,1,1,1,0,
                1,0,1,1,1,1,1,1,0,1,0,0,0,1,0,0,0,0,0,1,1,1,
                1,0,0,0,1,1,1,0,0,1,0,0,0,1,0,0,0,0,0,0,0,1,
                1,1,0,0,0,0,0,0,0,1,0,0,0,1,1,1,1,1,1,0,0,1,
                0,1,1,1,1,1,1,1,1,1,0,0,0,1,0,0,0,0,1,1,1,1,
                0,1,0,0,0,0,0,0,1,1,0,0,0,1,1,1,1,1,0,0,1,0,
                0,1,0,0,0,0,1,0,0,1,0,0,0,0,1,1,1,0,0,0,1,0,
                0,1,1,1,1,1,1,0,0,1,1,1,1,0,0,0,1,1,1,1,1,0,
                0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,1,0,0,0,0,0,
                1,1,1,1,1,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,
                1,1,0,0,1,1,1,0,0,0,1,0,0,1,0,0,0,0,0,1,1,0,
                1,1,0,0,1,0,0,0,1,1,1,0,0,1,1,1,1,0,0,1,1,0,
                0,1,1,1,1,1,0,0,1,0,1,1,1,1,1,1,1,1,1,1,1,0,
                0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
    ];


////sprite setup for cops and robbers
    //cops
    addSprite("https://i.imgur.com/lnVxiLL.png","cop");
    var copsprite= findSprite(sprites,"cop");
    //robber
    addSprite("https://i.imgur.com/4ixYlLR.png","robber");
    var robbersprite= findSprite(sprites,"robber");



    //cop setup
    copsprite.image.width= cellWidth;
    copsprite.image.height=cellHeight;
    var copArray=new Array();
    var numCops=3;
    var copCount = numCops;
    //var spawnDistance = 10;
    
    makeCops(numCops);
    
    
    
    function makeCops(numCops){
    
        for(var i=0; i<numCops; i++){

            //setup start position
            
            var startposX;
            var startposY;
            var walkable=false;
            while(!walkable){
                startposX= Math.floor(Math.random()*xcellCount);
                startposY= Math.floor( Math.random()*ycellCount);
                walkable= checkWalkable(startposX,startposY);
               
            }

            //console.log(startposX+" "+ startposY+""+walkable);
            addDrawnSprites(copsprite, startposX*cellWidth, startposY*cellHeight,copsprite.image.width, copsprite.image.height,"cop"+i);
            var thisCop=findSprite(drawnSprites,"cop"+i);
            copArray.push(thisCop);
            thisCop.posX=thisCop.X/cellWidth;
            thisCop.posY=thisCop.Y/cellHeight;
            thisCop.collider = makeSpriteCollider(thisCop, "box");


            if(thisCop.ID=="cop1"){
                thisCop.user=true;
            }

            thisCop.update=function(){
            
               if(this.user){
                   this.moved=false;
                   walkable=false;
                    if(direction=="up"){
                        walkable=checkWalkable(this.posX,this.posY-1);
                            if(walkable){
                                this.posY--;
                                this.moved=true;
                            }
                    }
                    else if(direction=="down"){
                        walkable=checkWalkable(this.posX,this.posY+1);
                        if(walkable){
                            this.posY++;
                            this.moved=true;
                        }
                    }
                    else if(direction=="left"){
                        walkable=checkWalkable(this.posX-1,this.posY);
                        if(walkable){
                            this.posX--;
                            this.moved=true;
                        }
                    }
                    else if(direction=="right"){
                        walkable=checkWalkable(this.posX+1,this.posY);    
                        if(walkable){
                            this.posX++;
                            this.moved=true;
                        }                  
                    }
                    direction="none";
                    if(turn >= maxTurns){
                        gameFinished = -1;
                    }
               }
               else{
                    if(robberArray.length>0){
                    var closestRobber=findClosest(this.posX, this.posY,robberArray);
                    }
                //  console.log(closestRobber.posX+ " "+ closestRobber.posY);

                // context.fillStyle="red";
                //context.fillRect(closestRobber.posX*cellWidth, closestRobber.posY*cellHeight,closestRobber.image.width,closestRobber.image.height);
                //copDirection=
                 if(closestRobber!=null){
                 var nextTile= directionByAstar(this.posX,this.posY,closestRobber.posX,closestRobber.posY,LevelGridArray);
                    if(nextTile!=null){ 
                        this.posX=nextTile.posX;
                        this.posY=nextTile.posY;
                        this.X=this.posX*cellWidth;
                        this.Y=this.posY*cellHeight;
                    }
                 }
                this.collider.update();
                this.collision();


            }
            this.X=this.posX*cellWidth;
            this.Y=this.posY*cellHeight;
            }

            thisCop.collision = function(){
            	for(var i = 0; i < robberArray.length; i++){
            		if(this.posX == robberArray[i].posX && this.posY == robberArray[i].posY){
            			robberArray[i].destroy();
            		}
            	}
            }
            
            thisCop.destroy = function(){
            	removeDrawnSprite(this);
            }

        }
    }  
    
//cop setup
robbersprite.image.width= cellWidth;
robbersprite.image.height=cellHeight;
var robberArray=new Array();
var numRobbers=2;
var robberCount = numRobbers;
var numCaughtRobbers=0;
//var spawnDistance = 10;

makeRobbers(numRobbers);



function makeRobbers(numRobbers){

    for(var i=0; i<numRobbers; i++){

        //setup start position
        
        var startposX;
        var startposY;
        var walkable=false;
        while(!walkable){
            startposX= Math.floor(Math.random()*xcellCount);
            startposY= Math.floor( Math.random()*ycellCount);
            walkable= checkWalkable(startposX,startposY);
           
        }

       // console.log(startposX+" "+ startposY+""+walkable);
       addDrawnSprites(robbersprite, startposX*cellWidth, startposY*cellHeight,robbersprite.image.width, robbersprite.image.height,"robber"+i);
       var thisRobber=findSprite(drawnSprites,"robber"+i);
       thisRobber.collider = makeSpriteCollider(thisRobber, "box");
       robberArray.push(thisRobber);
        thisRobber.posX=thisRobber.X/cellWidth;
        thisRobber.posY=thisRobber.Y/cellHeight;

    thisRobber.update= function(){
      
        var pposX=this.X/cellWidth;
        var pposY=this.Y/cellHeight;
        var moveable=false;
        var whilecounter=0;
        while(!moveable){
             pposX=this.X/cellWidth;
             pposY=this.Y/cellHeight;
            randomdirection=Math.floor(Math.random()*4);
            if(randomdirection<2){
                if(randomdirection<1){
                    pposY++;
                }
                else{
                    pposY--;
                }
            }
            else{
                if(randomdirection<3){
                    pposX++;
                }
                else{
                    pposX--;
                }

            }
            moveable= checkWalkable(pposX,pposY);
            var copisthere=checkCop(pposX,pposY);
            if(copisthere){
                moveable=false;
            }
           whilecounter++;
           if(whilecounter>24){
               moveable=true;
               var pposX=this.X/cellWidth;
               var pposY=this.Y/cellHeight;
           }
        }
       //var furthestTile= findFurthestSpot(this.posX, this.posY,copArray);
      // var nextTile= directionByAstar(this.posX,this.posY,furthestTile.posX,furthestTile.posY,LevelGridArray);

        this.posX=pposX;
        this.posY=pposY;
       //if(nextTile!=null){
            //this.posX=nextTile.posX;
           // this.posY=nextTile.posY;
            this.X = pposX*cellWidth;
            this.Y = pposY*cellHeight;
       //}
       // console.log(pposX+" "+pposY);

        this.collider.update();
        this.collision();
    }
    
    thisRobber.collision = function(){
    	/*console.log(this.collider.top + " robber " + this.collider.bottom);
    	for(var i=0; i<this.collider.collidedObjectsArray.length; i++){
    		console.log("aaa");
			if(this.collider.collidedObjectsArray[i] != null){
				console.log("fadsd");
		    	robberCount--;
		    	removeDrawnSprite(this);
		    	removeCollider(this.collider);
		    	if(robberCount == 0){
		    		finishGame();
		    	}

			}
		}*/
    	for(var i = 0; i < copArray.length; i++){
    		if(this.posX == copArray[i].posX && this.posY == copArray[i].posY){
                this.destroy();
                numCaughtRobbers++;
    		}
    	}
    }
    
    thisRobber.destroy = function(){
    	console.log(numRobbers);
		robberCount--;
		removeDrawnSprite(this);
		for(var j = 0; j < robberArray.length; j++){
			if(this.ID == robberArray[j].ID){
				robberArray.splice(j,1);
			}
		}
		if(robberArray.length == 0){
			gameFinished = 1;
		}
		
    }
    
    }
}  
    


function findClosest(copPosX,copPosY,thisArray){
    var closestRobber=thisArray[0];
    var crDistance=actualDistance(copPosX,copPosY,closestRobber.posX,closestRobber.posY);
    for(var i=1; i<thisArray.length; i++){
        var tempDistance=actualDistance(copPosX,copPosY,thisArray[i].posX,thisArray[i].posY);
        if(tempDistance<crDistance){
            crDistance=tempDistance;
            closestRobber=thisArray[i];
        }
    }
    return closestRobber;

}
function findFurthestSpot(robposX, robposY,copArray){
   var  closestCop= findClosest(robposX,robposY,copArray);
   var  tiles= loadtiles(LevelGridArray);
   var goal;
    for(var i=0; i<tiles.length; i++){
        if(closestCop.posX==tiles[i].posX && closestCop.posY==tiles[i].posY){
            goal=tiles[i];
            tiles[i].distanceToGoal=0;
        }  
    }
    var furthesttile=goal;
    for(var i=0; i<tiles.length; i++){
        tiles[i].calcDistanceToCity(goal);
        if(tiles[i].distanceToGoal>furthesttile.distanceToGoal){
            furthesttile= tiles[i];
        }
    }
    return furthesttile;
    
}

////////functions... to move to engine??
    function checkWalkable(posX,posY){
        if(posX>=xcellCount){
            return false;
        }
        if(LevelGridArray[posY*xcellCount +posX] !=null){
        if(LevelGridArray[posY*xcellCount +posX] ==1){
            return true;
        }
        else{
            return false;
        }
    }

    }
    function checkCop(posX,posY){
        var cop=false;
        for(var i=0; i<copArray.length; i++){
            if(copArray[i].posX==posX && copArray[i].posY==posY){
                cop=true;
            }
        }
        return cop;
    }
function finishGame(){
	while(copArray.length > 0){
		copArray[0].destroy();
		copArray.splice(0,1);
	}
	while(robberArray.length > 0){
		robberArray[0].destroy();
	}
	console.log(numRobbers);
	robberCount = numRobbers;
	makeCops(numCops);
    makeRobbers(numRobbers);
    turn = 0;
	numCaughtRobbers = 0;
	gameFinished = 0;
}



var turn=0;
var robberturn=0;
var copturn=0;
var waitnumber=6;
var waitcounter=0;

function update(){

    if(waitcounter>=6){
         if(robberturn<robberArray.length){
             robberArray[robberturn].update();
            robberturn++;
           // console.log(robberturn+"robberturn");
            waitcounter=0;
        }
        else if(copturn<copArray.length){
             copArray[copturn].update();
            if(copArray[copturn].ID=="cop1"){
                copArray[copturn].color="green";
                if (copArray[copturn].moved){
                    copArray[copturn].color="red";
                copturn++;
                waitcounter=0;
                }
                else{
                    waitcounter++
                }
        }
            else{
                copturn++;
                waitcounter=0;
            }
             //if user... then no update to this...
            //copturn++;
         //   console.log(copturn+"copturn");

            //if user... then waitcounter++ to keep open the turn
           // waitcounter=0;
         }
         else{
            turn++;
            robberturn=0;
            copturn=0;
          //  console.log(turn+ "turn");
            waitcounter=0;

        }
     
    }
    else{
        waitcounter++;
    }
   


    direction="none";
}




function draw(){
    
        canvas.width = canvas.width;
    
       
    
        context.fillStyle="grey";
        context.font = '16px monospace';
        context.fillRect(0,cellHeight*ycellCount,canvas.width,canvas.height-cellHeight*ycellCount);
        context.fillStyle="black";
       // context.strokeRect(0, cellHeight*ycellCount,canvas.width,canvas.height-cellHeight*ycellCount);
        //context.fillRect(0,0,canvas.width,cellHeight*ycellCount);       // drawStats();



        //grid drawing
        
       // context.fillRect(0,0,cellWidth,cellHeight);

      
    

        for(var i=0; i<LevelGridArray.length; i++){
            context.fillStyle="white";
           if(LevelGridArray[i]==1){
                    var ix = i-(Math.floor(i/(xcellCount))*xcellCount);
                    var iy =Math.floor(i/(xcellCount));
                    context.fillRect(ix*cellWidth,iy*cellHeight,cellWidth,cellHeight);
                    context.fillStyle="black";
                    context.strokeRect(ix*cellWidth,iy*cellHeight,cellWidth,cellHeight);
                  //  console.log(ix + " "+iy+ " " +iy*cellHeight);
           }

        }

        userSprite=findSprite(copArray,"cop1");
        context.fillStyle=userSprite.color;
        context.fillRect(userSprite.X,userSprite.Y,userSprite.image.width,userSprite.image.height);
        drawSprites();

        addText("30","comic Sans","Turns remaining:"+(maxTurns-turn),10,cellHeight*ycellCount+15,"white");
        addText("30","comic Sans","Robbers Caught:"+numCaughtRobbers,10,cellHeight*ycellCount+30,"white");
        
       /* if(onMenu){
            context.fillStyle = 'white';
            context.font = '48px monospace';
            measurement = context.measureText(title);
            context.fillText(title, (context.canvas.width - measurement.width) / 2, context.canvas.height / 2);
    
            context.fillStyle = 'red';
            context.font = '24px monospace';
            measurement = context.measureText(text);
            context.fillText(text, (context.canvas.width - measurement.width) / 2, context.canvas.height / 2 + 30);
        } */	
        if(gameFinished){
        	var text;
        	if(gameFinished == 1){
        		text = "Congratulations! You've won."
        	}
        	else{
        		text = "The robbers got away."
        	}
        	context.font = '50px monospace';
        	var textMeasurement = context.measureText(text);
        	var textXPos = (context.canvas.width - textMeasurement.width) / 2;
        	var textYPos = context.canvas.height / 2 - 50;
        	context.fillStyle = 'red';
    		context.fillText(text, textXPos, textYPos);
    		
    		var text2 = "Press spacebar to restart.";
    		var text2Measurement = context.measureText(text2);
    		var text2XPos = (context.canvas.width - text2Measurement.width) / 2;
    		context.fillText(text2, text2XPos, textYPos + 100)
    		
    		if(spacebar){
    			finishGame();
    		}
        }
    }
